Add favorite buildings

hannesmcman opened the pull request:
This should resolve #1960. I added a new component for the favorite button and I don't know if I followed best practices with storing which buildings were favorited, but I learned a lot in the process! I wasn't able to get the app running on an android simulator...kept getting this weird bug about multiple dex files.

hannesmcman committed:
added favorite feature to building hours

hannesmcman committed:
removed console.logs

hawkrives reviewed (commented):
I'll review in a few days with some actual comments, but thanks for the PR! I know our contributing guidelines aren't very useful for figuring out how the app works. 

My first comment: you might want to look at lib/storage.js and create a pair of functions to store the set of favorited buildings so that they're persisted across app restarts.

hannesmcman committed:
updated to fix problems with CI tests

hannesmcman committed:
joined if branches of FavoriteButton

hannesmcman committed:
added storage for favorite buildings

hannesmcman committed:
fixed all but one of codeclimate errors

hannesmcman committed:
fixed some codeclimate errors

hawkrives commented on source/flux/parts/buildings.js:
```diff
@@ -0,0 +1,44 @@
+
+type SaveFavoriteBuildingsAction = {
+  type: 'buildings/SAVE_FAVORITE_BUILDINGS',
+  payload: Array<BuildingType>,
```
It's probably significantly simpler overall to store an array of the building names as strings instead of the actual building information. Also requires less space on the device and doesn't raise so many issues of caching. 

Would you mind working this over to use an array of strings instead of BuildingType, please?

hawkrives replied:
Well, so much for the "start a review" button… Review in progress!

hawkrives commented on source/lib/storage.js:
```diff
@@ -74,3 +75,13 @@ export function setCredentialsValid(valid: boolean) {
+export function setFavoriteBuildings(buildings: Array<BuildingType>) {
+  return setItem(favoriteBuildingsKey, buildings)
+}
+export function getFavoriteBuildings() {
```
I'd like a return type on this function, please; probably `Promise<Array<BuildingType>>` given this code. It should return the same type as what was stored, essentially.

hawkrives commented on source/views/building-hours/detail/index.js:
```diff
@@ -11,17 +11,45 @@ import moment from 'moment-timezone'
+        <FavoriteButton
+          favorited={favoriteBuildings.includes(building)}
+          navigation={navigation}
+          onFavorite={() => {
```
I'd like to pull this logic out of here, probably into a function in parts/building-hours.js? Said function would be called as part of the saveBuildingFavorites function.

hawkrives replied:
Essentially, then, we avoid mutating the array here entirely and call onSaveBuildingFavorites or whatever it's called (sorry, reviewing from my phone atm) instead. Redux will cause a re-render or the screen when the list of favorited buildings changes, so we'll know that the building is favorited because it suddenly will be (if that makes sense; I can try to explain better if you want me to.)

hawkrives replied:
You'll also need to add the dispatch call / map thing in here, instead of in StatefulList, because that's how this component will have access to the function.

hawkrives commented on source/views/building-hours/stateful-list.js:
```diff
@@ -27,7 +29,31 @@ const groupBuildings = (buildings: BuildingType[]) => {
 }
 
-type Props = TopLevelViewPropsType
+const loadAllBuildings = (
```
I'd like to see this logic incorporated into groupBuildings itself, if you don't mind. I can show you how if you'd like.

hawkrives commented on source/views/building-hours/stateful-list.js:
```diff
@@ -48,16 +74,15 @@ export class BuildingHoursView extends React.Component<Props, State> {
     // This updates the screen every ten seconds, so that the building
     // info statuses are updated without needing to leave and come back.
-    this.setState({intervalId: setInterval(this.updateTime, 1000)})
+    this.setState({intervalId: setInterval(this.updateTime, 10000)})
```
This change should be reverted, please.

drewvolz replied:
I think he was going off of the comment that says "ten seconds" so let's change the comment if we're going to leave the code at one second.

hawkrives replied:
Yes, we should change the comment.

hawkrives commented on source/views/building-hours/stateful-list.js:
```diff
@@ -92,24 +117,51 @@ export class BuildingHoursView extends React.Component<Props, State> {
 
+  onUpdate = () => {
+    this.fetchData()
+    this.props.onSaveFavorites(this.props.favoriteBuildings)
```
This is the other half of my comment about moving the "make a favorite" logic into the redux file; when you do that, you won't need to save in onUpdate at all, because the update will have happened as soon as the heart button was pressed.

hawkrives commented on source/views/building-hours/stateful-list.js:
```diff
@@ -92,24 +117,51 @@ export class BuildingHoursView extends React.Component<Props, State> {
+  }
+}
+
+function mapDispatch(dispatch): ReduxDispatchProps {
```
This one will need to move in to the Detail screen, because we'll be calling onSaveFavorites from there.

hawkrives commented on source/views/components/nav-buttons/favorite.js:
```diff
@@ -0,0 +1,55 @@
+}
+
+export class FavoriteButton extends React.PureComponent<Props, State> {
+  state = {
```
When you move the logic of saving the favorite into the redux file and make the "am I favorited" logic on the Detail screen be a (mathematical) function of the Redux state, this component won't need any internal state any longer; the lifecycle will go something like this:

- Person taps "favorite"
- Redux "action creator" onSaveFavorites is called
- Said function determines an array of actually favorited buildings (and saves it to disk)
- Created action is dispatched via the `connect`ed dispatch function
- Redux state updates with new array of favorites building names
- Redux state update causes React to re-render the components that have connected to the store
- StatefulList gets the new set of building names and groups them again
- DetailView gets the new set of building names and figures out if it's favorited
- If it is, it sets the "favorited" prop of FavoriteButton to "true" or "false", and react calls FavoriteButton's render function again with the new props (otherwise nothing happens)

At that point, there's no reason to keep the separate state in this button, so it becomes a much smaller component. (Not that it's not already small.)

If you try that and the button doesn't seem to be working, please let me know - the router that handles new screens might be interfering with the props getting passed right (it may not be passing the new ones, since navigate is only called once and we haven't changed screens). Or it may not! I'm not sure.

hannesmcman replied:
Ok, I have implemented the changes you asked for, but I am unsure how to get the navigationOptions in the detail view to refresh when the view is re-rendered by redux upon the state change. The only way I can get the button to update properly is if I pass navigation to the onSaveFavorites and saveFavoriteBuildings functions, respectively and then call navigation.setParams({favoriteBuildings: favoriteBuildings}) within saveFavoriteBuildings in buildings.js...is there a way to get the navigationOptions to automatically refresh when the redux state changes?

hannesmcman replied:
@hawkrives Regarding "npm run check"..running yarn does not make the "check" script work..

hawkrives replied:
Sigh… that's what I was afraid of, that the toolbar buttons wouldn't be able to subscribe to redux. 

I'll think about this and probably get back to you tomorrow … I'm thinking that it might be possible / simple to connect to redux on the FavoriteButton itself, and pass the building name to the button, then use the buildings name in mapStateToProps within the button's redux connection to determine if the building is connected or not. 

I'll try to work up something for that approach tomorrow.

hannesmcman replied:
Sounds good. I just pushed the latest version that implemented most of the changes you requested.

hawkrives commented on source/views/building-hours/stateful-list.js:
```diff
@@ -92,24 +117,51 @@ export class BuildingHoursView extends React.Component<Props, State> {
     if (this.state.error) {
       return <NoticeView text={`Error: ${this.state.error.message}`} />
     }
-
```
Please put this space back in :)

hawkrives commented on source/views/building-hours/stateful-list.js:
```diff
@@ -48,16 +74,15 @@ export class BuildingHoursView extends React.Component<Props, State> {
 
   componentWillMount() {
     this.fetchData()
-
```
Please put this space back :)

hawkrives commented on source/views/building-hours/detail/index.js:
```diff
@@ -37,8 +65,11 @@ export class BuildingHoursDetailView extends React.PureComponent<Props, State> {
     this.setState({intervalId: setInterval(this.updateTime, 1000)})
   }
 
+  componentDidMount() {}
```
You can probably remove this method; I don't think you'll need it here.

hawkrives commented on ios/AllAboutOlaf/Info.plist:
```diff
@@ -76,7 +76,7 @@
 	<string>We use your calendar to add events to your calendar so that you remember what you wanted to attend.</string>
 	<key>NSLocationWhenInUseUsageDescription</key>
-	<string>We only use your location when you would expect us to, and never when you aren&apos;t using the app.</string>
+	<string>We only use your location when you would expect us to, and never when you aren't using the app.</string>
```
Please revert this (Xcode changes it, and something else changes it back… it's a struggle)

hawkrives reviewed (changes requested):
@hannesmcman Wow, thanks!

I left several comments and one novel this time around; I think you can push this through in the next few days (if you need help or have questions with it, please let us know; we know that the docs for how this ball of mud works essentially don't exist).

But overall, good work!

drewvolz commented on source/views/building-hours/stateful-list.js:
```diff
@@ -27,7 +29,31 @@ const groupBuildings = (buildings: BuildingType[]) => {
+  let allBuildings = Array()
+  allBuildings.push({
+    title: 'Favorites',
+    data: favoriteBuildings,
```
I'd like to see logic incorporated that restricted the 'Favorites' section from appearing if there are no favorites, if you don't mind. If you don't want to implement this, I'd be happy to work on it.

hannesmcman replied:
https://github.com/StoDevX/AAO-React-Native/pull/2003/commits/3cdeb6c37bed1768cc0ad1a0e5f0ed6d74cf52d8 
I did it the easy way in this commit, but if you think there's a better way you should go for it. 👍

hannesmcman committed:
updated logic in response to review

hannesmcman committed:
removed favorites label from list if there are no favorites

hawkrives commented on source/views/building-hours/detail/index.js:
```diff
@@ -98,10 +107,22 @@ export class BuildingHoursDetailView extends React.PureComponent<Props, State> {
 function mapStateToProps(state: ReduxState): ReduxStateProps {
   return {
-    favoriteBuildings: state.buildings ? state.buildings.favoriteBuildings : [],
+    favoriteBuildings: state.buildings.favoriteBuildings
```
We know that if state.buildings exists, then favoriteBuildings will too; it's the buildings key itself that may not exist (this line was correct before [unless something was yelling about it in which c add I am probably wrong])

hawkrives commented on source/views/building-hours/stateful-list.js:
```diff
@@ -92,19 +117,21 @@ export class BuildingHoursView extends React.Component<Props, State> {
   }
 
   render() {
+    this.fetchData()
```
Why did you add this call here?

hannesmcman replied:
Without it, the favorites are not added in the re-render of stateful-list, because groupBuildings is not called.

hawkrives committed:
remove empty anonymous function call from FavoriteButton

If you just pass a reference to a function without calling it, it's synonymous to passing an
anonymous function that passes no arguments.

That is, `prop={() => {func()}}` is nearly equivalent to `prop={func}` for almost all intents
and purposes.

hawkrives committed:
reduce duplicated code in picking the icon name to render

hawkrives committed:
update the favoriteBuildings check in stateful-list

If `state.buildings` exists, `state.buildings.favoriteBuildings` is guaranteed to exist.

We need to check if `state.buildings` itself exists, so that's what this commit does.

hawkrives committed:
switch to functional array manipulation in groupBuildings

We avoid manual array manipulation in the app, in favor of the .map/.reduce functions.

You can see how the first loop, which builds up an array of favorited buildings, can be
replaced with a single line Array#filter call that "filters" down the existing array to just
the items we're interested in. I also merged the array creation with the group creation on
the previous line.

Then, later, we had some array manipulation to build up the total set of groups to show; I
replaced the if-and-loop with a single if that contains an "array spread" to prepend the
Favorites group if we need to.

hawkrives committed:
remove fetchData call from StatefulList.render

hawkrives committed:
add a component that connects FavoriteButton to redux

Adding this as a separate component allows us to potentially re-use the "favorite" component
somewhere else later on, where we'll only need to re-connect it to redux or wherever the data
comes from.

hawkrives committed:
remove SAVE_FAVORITE_BUILDINGS action in favor of TOGGLE_FAVORITE_BUILDING and LOAD_FAVORITE_BUILDINGS

LOAD_FAVORITE_BUILDINGS is a single-purpose action that loads the favorites from AsyncStorage
and tells the reducer to store them.

TOGGLE_FAVORITE_BUILDING is a much fancier action creator, that gets the current set of
favorited buildings from the redux state and uses that to add or remove the given building
name from the set of favorites, and then re-sort the list of buildings.

It also has to save the favorites to AsyncStorage, because there's no better place to perform
the actual "save" (yet).

hawkrives committed:
shrink the giant comment at the top of BuildingHoursDetail

hawkrives committed:
remove need to connect the BuildingHours detail view to redux

Now that we have the redux-connected ConnectedBuildingFavoriteButton component, nothing within
the detail screen actually needs Redux connectivity. So let's remove it!

hawkrives committed:
rename "buildings.favoriteBuildings" to "buildings.favorites"

hawkrives committed:
prettify

hawkrives committed:
remove giant comment from top of stateful-list

hawkrives committed:
stop passing favoriteBuildings to the plain building list

It was only used to be passed to the detail view, but the detail view no
longer needs it either.

hawkrives committed:
group buildings at render time

This avoids the need to cover all edge cases about when data can update,
and I'm ok with it because groupBuildings isn't a terrible expensive
function.

hawkrives committed:
Merge branch 'master' into add-favorite-buildings-notes

* master: (31 commits)
  chore(package): update lockfile
  chore(package): update danger to version 2.1.7
  chore(package): update lockfile
  chore(package): update babel-eslint to version 8.1.2
  Remove badges from README.md
  chore(package): update lockfile
  chore(package): update babel-eslint to version 8.1.1
  chore(package): update lockfile
  chore(package): update babel-eslint to version 8.1.0
  chore(package): update lockfile
  chore(package): update eslint to version 4.14.0
  chore(package): update lockfile
  fix(package): update glamorous-native to version 1.2.1
  chore(package): update lockfile
  chore(package): update jest to version 22.0.4
  chore(package): update lockfile
  chore(package): update babel-jest to version 22.0.4
  notice of break hours
  hide pause menu b/c break
  chore(package): update lockfile
  ...

hawkrives committed:
resolve inverted favorite icon

hawkrives committed:
Revert provisioning changes from 2933af16b614a454bbdf18ce3745ce1a169e3619

hawkrives committed:
replace "return" with "dispatch" inside action creator thunk

hawkrives committed:
remove now-unused "favorited" key from BuildingType

hawkrives committed:
call groupBuildings at the appropriate points

so that we don't have to call it on every render… I wasn't going to
write this code, but then I wrote it to test something, and now I feel
bad not using it, as it's somewhat more efficient than the other.

hawkrives committed:
collapse an object

hawkrives reviewed (approved):
@hannesmcman I think this is ready to merge and ship as part of 2.5! Great work on spearheading this!

hawkrives committed:
run prettier

hawkrives committed:
sort lifecycle methods as requested by eslint
